### PR TITLE
Nkai priority rework

### PR DESCRIPTION
**Pull request by nullkiller**
The goal of this PR is to improve defense. AI will attack enemy heroes around their towns more aggressively.
Also reworked priority engine a bit. Movement spending is not considered now as a bad effect but instead it modulates value of other gains. Added army growth as a separate fuzzy variable to not clash with one-time army value.

### CODE REVIEW

**Comment by nullkiller** [comment]
Work is still in progress but in overall it looks playable already

**Comment by DjWarmonger** [comment]
I gave it a try, AI finished 4-player game on my test Jebus XL+U under 5 months :) No town ping-pong for sure.

File: AI/Nullkiller/Markers/DefendTown.h [hunk 1]
```diff
@@ -37,6 +38,8 @@ namespace Goals
 		uint64_t getDefenceStrength() const { return defenceArmyStrength; }
 
 		uint8_t getTurn() const { return turn; }
+
+		bool isContrAttack() { return contrattack; }
```
Comment (IvanSavenko) [inline AI/Nullkiller/Markers/DefendTown.h:42]:
```suggestion
		bool isCounterAttack() { return counterattack; }
```

File: AI/Nullkiller/Analyzers/DangerHitMapAnalyzer.cpp [hunk 2]
```diff
@@ -67,34 +76,53 @@ void DangerHitMapAnalyzer::updateHitMap()
...
 					auto objects = cb->getVisitableObjs(pos, false);
 					
 					for(auto obj : objects)
 					{
 						if(cb->getPlayerRelations(obj->tempOwner, ai->playerID) != PlayerRelations::ENEMIES)
 							enemyHeroAccessibleObjects[path.targetHero].insert(obj);
+
+						if(obj->ID == Obj::TOWN && obj->getOwner() == ai->playerID)
+						{
+							auto & treats = townTreats[obj->id];
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply AI/Nullkiller/Analyzers/DangerHitMapAnalyzer.cpp:109]: Wow, really. Will fix

File: AI/Nullkiller/Analyzers/DangerHitMapAnalyzer.h [hunk 3]
```diff
@@ -51,23 +55,34 @@ struct HitMapNode
 	}
 };
 
+struct TileOwner
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply AI/Nullkiller/Analyzers/DangerHitMapAnalyzer.h:58]: Will check

File: AI/Nullkiller/Engine/PriorityEvaluator.cpp [hunk 4]
```diff
@@ -407,13 +479,22 @@ float RewardEvaluator::getStrategicalValue(const CGObjectInstance * target) cons
...
+
+			return std::min(1.0f, std::sqrt(armyIncome / 40000.0f)) + std::min(0.3f, dailyIncome / 10000.0f);
+		}
+
 		auto fortLevel = town->fortLevel();
-		auto booster = isAnotherAi(town, *ai->cb) ? 0.3 : 1;
+		auto booster = isAnotherAi(town, *ai->cb) ? 0.3f : 0.7f;
 
-		if(town->hasCapitol()) return 1;
+		if(town->hasCapitol()) return booster;
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply AI/Nullkiller/Engine/PriorityEvaluator.cpp:497]: I do not think this way it is prohibited by our code style. Does it?

[the remark this answers is not in the document]
Comment (IvanSavenko) [reply AI/Nullkiller/Engine/PriorityEvaluator.cpp:497]: Don't see it as actually listed as "prohibited" but not listed as allowed either & mostly not used in code, only in some ancient parts.

**Comment by dydzio0614** [comment]
I managed to get freeze on purple AI turn (applied warmonger's movement points compilation fixes to your branch) - save here: https://drive.google.com/file/d/182DmeM1VhbY63FtH_c_g24PGnziZt402/view?usp=sharing

**Comment by DjWarmonger** [comment]
Still, when AI loses hero, it tries to access it anyway and never forgets about defeated hero.
```
Cannot get object with id 5152. Object was removed
Accessing no longer accessible hero Christian!
Cannot get object with id 5152. Object was removed
Accessing no longer accessible hero Christian!
Cannot get object with id 5152. Object was removed
Accessing no longer accessible hero Christian!
Cannot get object with id 5152. Object was removed
Accessing no longer accessible hero Christian!
Cannot get object with id 5152. Object was removed
Accessing no longer accessible hero Christian!
Cannot get object with id 5152. Object was removed
Accessing no longer accessible hero Christian!
Hero Calh has 1630 MP left
Hero Thyroc has 1560 MP left
Player 0 (red) ends turn
Player 0 (red) ended turn
Player 1 (blue) starting turn, day 82
Cannot get object with id 5150. Object was removed
Accessing no longer accessible hero Ryland!
Cannot get object with id 5150. Object was removed
Accessing no longer accessible hero Ryland!
Cannot get object with id 5150. Object was removed
Accessing no longer accessible hero Ryland!
Cannot get object with id 5150. Object was removed
Accessing no longer accessible hero Ryland!
Player 1 (blue) ends turn
```
This should be eventually fixed after all thes years.

I observe that in such case AI stops using most of its heroes, they just stand still and pass turns.

**Comment by DjWarmonger** [comment]
In general, NKAI heroes tend to stand still even if there are some dwellings on Windmills around. That's why VCAI has "wander" feature, which makes bored heroes just explore nearby objects. Even if they are modded objects with no value known to AI.

**Comment by DjWarmonger** [comment]
This map featurers long paths (especially for red) and no unguarded treasures, though overall its pretty rich. I noticed that AI struggles in these conditions.
[144 OceansEleven2 20140402.zip](https://github.com/vcmi/vcmi/files/11998710/144.OceansEleven2.20140402.zip)

**Comment by DjWarmonger** [comment]
@nullkiller Any plans for improvements, or should we merge it as it is for 1.3?

**Comment by dydzio0614** [comment]
I would not merge it for 1.3 until it is fully ready

**Comment by DjWarmonger** [comment]
Well, it certainly works better than the previous version. Just still see a area for improvement.

**Comment by nullkiller** [comment]
No, I do not see it being a part of 1.3 as it is scheduled to be released soon. But I plan to start working on it again next week

**Comment by DjWarmonger** [comment]
I alrerady noticed that AI visits already cleared branks that give creatures as a reward.

**Comment by tianmaster** [comment]
Testing has found that if there are garrisoned heroes in an AI town and the hero has troops, other AI heroes visiting that town will cause an infinite loop

**Comment by nullkiller** [comment]
Thank you all. Will take a look onto it.

**Comment by DjWarmonger** [comment]
ALso, did you consider an intermediate input-output variable for hero and turnDistance, such as:

```
if heroRole is SCOUT and scoutTurnDistance is LOW then heroUtility is HIGH
if goldReward is HUGE and goldPreasure is HIGH then goldGain is HIGHEST
if heroUtility is HIGH and goldGain is HIGH then Value is HIGHEST
```

This might make it more readable and also faster to process since there will be fewer short rules?

**Comment by DjWarmonger** [comment]
> I tried current version, noticed that AI re-visits already cleared banks that give creatures as a reward.

Unfortunately this issue is now present in current beta as well, needs a fix :?

**Comment by DjWarmonger** [comment]
After a few games I can say this version of AI is a significant step forward. It not only advances faster in general, but it's quite robust and doesn't get stuck and continues to play even in difficult circumstances. No crashes, either.

So yes, I'd like to merge it for 1.3 and see what players say about it.

**Comment by misiokles** [comment]
Or maybe it's better to add NKAI-1.3 as choosable option in launcher?

**Comment by peterpaulis** [comment]
The AI in 1.2 is one of its weakest points, any improvement on this is a must in 1.3 imho

**Comment by nullkiller** [comment]
> Also, did you consider an intermediate input-output variable for hero and turnDistance, such as:
> 
> ```
> if heroRole is SCOUT and scoutTurnDistance is LOW then heroUtility is HIGH
> if goldReward is HUGE and goldPreasure is HIGH then goldGain is HIGHEST
> if heroUtility is HIGH and goldGain is HIGH then Value is HIGHEST
> ```
> 
> This might make it more readable and also faster to process since there will be fewer short rules?
> 
> Here is the trivial example: https://github.com/fuzzylite/fuzzylite/blob/release/examples/mamdani/SimpleDimmerChained.fll

Considered but not sure. Maybe in order to remove some duplicating logic it may be interesting so. What I really do not like is matrix when you have turnDistance HIGH LOW MEDIUM and for instance armyReward HIGH LOW MEDIUM it results in 9 rules to make full coverage

**Comment by nullkiller** [comment]
For now I detected a few freezes so it is still far from stable.

**Comment by DjWarmonger** [comment]
AI doesn't seem to recruit a hero once all its heroes have been defeated, despite still having 3 towns.

**Comment by nullkiller** [comment]
I addressed heroes recruiting a bit. If it will be not enough will fix it later. For now it looks playable.

**Comment by nullkiller** [comment]
It is still poorly tested so do not want to merge it into 1.3

**Comment by IvanSavenko** [comment]
What about 1.3.1/1.3.2?
Should give enough time to test this & players won't have to wait 3-4 months for these changes.

**Comment by DjWarmonger** [comment]
I think that adding brand new AI to 1.3.1 will cause more confusion than just 1.3.0. 1.3.1 is supposed to be a hotfix for issues found on the release.
So either push new AI for 1.3.0 to find and fix teh issues until 1.3.1, or wait for 1.3.2 or next.

Still, 1.3.2 is targeted at modders who may want to use extra features to prepare new mods in the meantime. It's not aimed at casual players.

**Comment by nullkiller** [comment]
Up to you. This branch is on top of beta. I continue testing.

**Comment by IvanSavenko** [comment]
In that case I'd rather merge it on 1.3.0/beta tomorrow & then fix any discovered issues for 1.3.1. Even if there are some issues this is still better than effectively postponing new features for 3-4 month till 1.4

Is everybody OK with that?

**Comment by nullkiller** [comment]
A set of freezes were addressed so it looks playable now. Still I expect something to pop out later

**Comment by dydzio0614** [comment]
I guess it sounds fair, considering that for people who do not struggle with technical/resolution problems and do not play on mobile or focus on campaigns, the 1.3 release is not very interesting. Sure, ability to zoom map etc. is nice, some mechanics bug fixes as well but it doesn't give new motivation for gameplay compared to what 1.2 offers. Notable AI improvements would be definitely game changer when it comes to singleplayer experience.

**Pull request merged**